File: content/problems/0039-combination-sum/solutions/sorted-backtracking/Solution.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> answer = new ArrayList<>();
        search(candidates, target, 0, new ArrayList<>(), answer);
        return answer;
    }

    private void search(int[] candidates, int remaining, int start, List<Integer> path, List<List<Integer>> answer) {
        if (remaining == 0) {
            answer.add(new ArrayList<>(path));
            return;
        }
        for (int index = start; index < candidates.length && candidates[index] <= remaining; index++) {
            path.add(candidates[index]);
            search(candidates, remaining - candidates[index], index, path, answer);
            path.remove(path.size() - 1);
        }
    }
}

File: content/problems/0039-combination-sum/solutions/sorted-backtracking/solution.cpp
class Solution {
    void search(const vector<int>& candidates, int remaining, int start, vector<int>& path, vector<vector<int>>& answer) {
        if (remaining == 0) {
            answer.push_back(path);
            return;
        }
        for (int index = start; index < (int)candidates.size() && candidates[index] <= remaining; ++index) {
            path.push_back(candidates[index]);
            search(candidates, remaining - candidates[index], index, path, answer);
            path.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> answer;
        vector<int> path;
        search(candidates, target, 0, path, answer);
        return answer;
    }
};
